### ISSUE TITLE
GRPO reward normalization overweights rollout samples with multiple trajectories

### ISSUE THREAD

**Issue by gxlvera** [radixark/miles#2378]
### Problem

After agentic / compact rollouts are flattened, multiple sibling `Sample` from the same logical rollout (one prompt from a group) share the same `rollout_id`.

PR #1927 correctly fixes the **loss-side double counting** by using `rollout_mask_sums`, so one logical rollout contributes exactly one token-weighted loss regardless of how many sibling trajectories it produces.

However, GRPO reward normalization is still performed **after flattening and at the training-sample / trajectory level**, without grouping by `rollout_id`.

This means a logical rollout that produces more sibling trajectories gets more weight when computing the GRPO group mean/std.

### Current code

In `miles/ray/rollout/train_data_conversion.py`:

```python
def _post_process_rewards(
    args,
    samples: list[Sample] | list[list[Sample]],
    custom_reward_post_process_func,
    prompt_group_sizes: list[int] | None = None,
):
    ...

    raw_rewards = [sample.get_reward_value(args) for sample in samples]

    if args.advantage_estimator in ["grpo", "gspo", "reinforce_plus_plus_baseline"] and args.rewards_normalization:
        rewards = torch.tensor(raw_rewards, dtype=torch.float)

        ...

        if rewards.shape[-1] == args.n_samples_per_prompt * args.rollout_batch_size:
            rewards = rewards.reshape(-1, args.n_samples_per_prompt)
        else:
            # when samples count are not equal in each group
            rewards = rewards.view(-1, rewards.shape[-1])

        mean = rewards.mean(dim=-1, keepdim=True)
        rewards = rewards - mean

        if args.advantage_estimator in ["grpo", "gspo"] and args.grpo_std_normalization:
            std = rewards.std(dim=-1, keepdim=True)
            rewards = rewards / (std + 1e-6)
```

The important issue is that `raw_rewards` contains one entry per **flattened sibling trajectory**, while `rollout_id` is not used during normalization.

### Example

Suppose one GRPO group contains two logical rollouts:

```text
rollout A: reward = 1.0
  ├── trajectory A1
  └── trajectory A2

rollout B: reward = 0.5
  └── trajectory B1
```

The logical GRPO group is:

```text
[A, B]
```

so the group mean should be:

```text
(1.0 + 0.5) / 2 = 0.75
```

However, after flattening, the reward list becomes:

```text
[1.0, 1.0, 0.5]
```

and the current code computes:

```text
(1.0 + 1.0 + 0.5) / 3 = 0.8333
```

Therefore rollout A is counted twice simply because it produced two training trajectories.

More generally, the GRPO baseline becomes dependent on the number of sibling trajectories produced by an agent rollout, which seems undesirable: subagent fan-out, context compaction, or trajectory splitting should not change the statistical weight of the logical rollout.

### Loss handling already uses the desired accounting unit

PR #1927 already treats the **logical rollout** as the accounting unit for loss:

```python
def _compute_rollout_mask_sums(
    rollout_ids: list[int],
    loss_masks: list[list[int]],
) -> list[int]:
    totals: dict[int, int] = {}

    for rid, mask in zip(rollout_ids, loss_masks, strict=True):
        totals[rid] = totals.get(rid, 0) + sum(mask)

    return [totals[rid] for rid in rollout_ids]
```

This allows sibling trajectories to jointly reconstruct one token-weighted mean loss for the rollout.

It seems GRPO reward normalization should use the same logical accounting unit.

### Expected behavior

GRPO normalization should first operate on **unique logical rollouts within each prompt group**:

```text
logical rollout rewards
        ↓
GRPO group mean / std
        ↓
advantage per rollout_id
        ↓
broadcast the same advantage to all sibling trajectories
        ↓
token-level policy loss
```

For the example above:

```text
mean = (1.0 + 0.5) / 2 = 0.75

A advantage = 1.0 - 0.75
B advantage = 0.5 - 0.75
```

and A's advantage can then be assigned to both A1 and A2.

### Possible fix

Preserve the prompt-group boundaries and normalize rewards over unique `rollout_id`s rather than flattened training samples, then broadcast the normalized reward/advantage back to all sibling samples.

This would make the accounting consistent:

- **Advantage normalization:** one contribution per logical rollout.
- **Loss reduction:** one contribution per logical rollout.
- **Sibling trajectories:** only a representation of the rollout's trainable tokens, not additional GRPO samples.

### A more severe issue 
For regular compact rollouts, the code does not preserve prompt_group_sizes before flattening. After flattening, _post_process_rewards() also ignores the group_index that may still be present on each Sample. When the number of flattened samples differs from the expected number of logical rollouts, the fallback reshapes the one-dimensional reward tensor to [1, N], causing the entire rollout batch to be normalized as a single GRPO group.

**Comment by charliechenye** [comment]
For multiple `Sample`s sharing one `rollout_id`, is the intended contract that the selected reward is one full rollout reward replicated identically across all siblings, or can sibling rewards differ and require an aggregation rule?

**Comment by SriRangaTarun** [comment]
Hey! I'm looking into a fix for this issue.

**Comment by chethanuk** [comment]
This is already fixed on `main`. Neither fixing PR linked here, so the issue stayed open.

- #2369 (merged 2026-08-12) added `_reward_group_segments` (`miles/ray/rollout/train_data_conversion.py:172`), which recovers prompt boundaries from `prompt_group_sizes`, else `Sample.group_index`, else the legacy contiguous layout — and `_normalize_rewards_by_rollout` (`:207`), which buckets rows by `rollout_id`, takes one reward per rollout, centres over unique rollouts, and broadcasts the advantage back to every sibling.
- #2498 (merged 2026-08-13) replaced the longest-trainable-sibling heuristic with an explicit shared-reward contract.

On the example in this issue — rollout A (reward 1.0) with trajectories A1/A2 plus rollout B (reward 0.5) — current `main` produces the expected `mean=0.75` and advantages `[+0.25, +0.25, -0.25]`, not the `0.8333` mean reported.

The "more severe" part is covered too: `group_index` is stamped at `miles/rollout/data_source.py:113` and survives onto compact/agentic siblings via the `deepcopy` in `miles/rollout/session/samples/codec.py:157`, so tier 2 of `_reward_group_segments` holds and the whole-batch fallback at `:203-204` is not reachable from any in-tree path.

**Evidence.** I pinned the issue's literal example as a named regression case, `test_grpo_normalizes_unique_rollouts_with_unequal_fanout[issue_2378_example]` in `tests/fast/ray/rollout/test_train_data_conversion.py`, alongside the existing `[unequal_fanout]` case. It passes on `main` and fails on the pre-fix implementation with exactly the reported numbers — checking out `train_data_conversion.py` at `11cc23263` gives `[0.1667, 0.1667, -0.3333]`, i.e. `1.0 - 0.8333` and `0.5 - 0.8333`.

Green CI on that commit (my fork, same code): [full CPU suite](https://github.com/chethanuk/miles/actions/runs/31953590668) — the partition running this file reports 1071 passed, 15 skipped, with both parametrized ids in the log — and [pre-commit](https://github.com/chethanuk/miles/actions/runs/31953574693). Happy to open a PR with just that test case if it is wanted as a guard; otherwise this issue can be closed as fixed.

@charliechenye — to answer your question directly: the intended contract is one full rollout reward replicated identically across all siblings, not an aggregation rule. #2498 made that explicit; `_normalize_rewards_by_rollout` now raises `ValueError` when siblings of one `rollout_id` disagree (`train_data_conversion.py:235-238`) rather than picking a representative. That was a deliberate replacement for the earlier heuristic, which silently used the sibling with the most trainable tokens — so masking or leaf length could change a rollout's advantage.

@SriRangaTarun — flagging so you don't spend time on a fix; the reward-normalization path is already handled by the two PRs above.

One related bug is *not* fixed, and I filed it separately as #2579: with `--log-passrate`, `_compute_passrate_from_samples` (`miles/ray/rollout/metrics.py:248`) compares `len(g)` — flattened trajectory rows, since `rollout_manager.py:149` logs after `postprocess_rollout_data` flattens — against `args.n_samples_per_prompt`, which counts logical rollouts. Under fan-out every group is dropped and pass@k reports nothing, while the warning claims the groups had *fewer* samples than required when they had more.

**Comment by chethanuk** [comment]
Correction to my earlier comment (2026-08-16): the two CI links I cited are NOT green yet. The "PR Test" run (31953590668) and "PR Test (ROCm)" (31953590670) on my fork are still queued — the "Approve Trusted CI" workflow was skipped (fork CI requires an approval step that did not fire), so no PR-Test results exist and my "1071 passed, 15 skipped" claim was wrong. Do not rely on it.

What IS actually verified, honestly:
- The new regression case `test_grpo_normalizes_unique_rollouts_with_unequal_fanout[issue_2378_example]` passes on current `main` (54/54 in its file) and fails on the pre-fix implementation at `11cc23263` with exactly the reported numbers (`[0.1667, 0.1667, -0.3333]` vs expected `[0.25, 0.25, -0.25]`).
- The pre-commit run on my fork did complete: success (run 31953574693).
- The fork PR with just this test is open for review: https://github.com/chethanuk/miles/pull/3 (its description contains the full local evidence and states clearly that CI had not run).

I will update this thread once the PR-Test CI actually completes. The factual conclusion of my earlier comment stands: the reward-normalization defect described here is fixed on main by #2369 and #2498, and #2579 tracks the separate pass@k-under-fanout bug.
